File: LeetCode Solutions/Problems/1171. Remove Zero Sum Consecutive Nodes from Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeZeroSumSublists(self, head: Optional[ListNode]) -> Optional[ListNode]:
        node = ListNode(0)
        node.next = head
        d, dv = {}, []
        s, nodeV = 0, node
        while nodeV:
            s += nodeV.val
            if s not in d:
                d[s] = nodeV
                dv.append(s)
            else:
                d[s].next = nodeV.next
                while dv[-1] != s:
                    t = dv.pop()
                    del d[t]
            nodeV = nodeV.next
        return node.next

File: LeetCode Solutions/Problems/1171. Remove Zero Sum Consecutive Nodes from Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeZeroSumSublists(ListNode* head) {
        ListNode node(0);
        ListNode* nodeV = &node;
        nodeV->next = head;
        ListNode* headA = nodeV;
        unordered_map<int, ListNode* > d;
        vector<int> dv;
        int s = 0;
        while (nodeV){
            s += nodeV->val;
            if (d.count(s) == 0){
                d[s] = nodeV;
                dv.push_back(s);
            } else {
                d[s]->next = nodeV->next;
                while (dv.back() != s){
                    int t = dv.back();
                    dv.pop_back();
                    d.erase(t);
                }
            }
            nodeV = nodeV->next;
        }
        return headA->next;
    }
};